file.c

```diff
@@ -0,0 +1,289 @@
+    f->olen = 0;
+    errno = 0;
+}
+
+size_t fileread(mfile f, void *buf, size_t len, int64_t deadline) {
+    /* If there's enough data in the buffer it's easy. */
+    if(f->ilen >= len) {
+        memcpy(buf, &f->ibuf[f->ifirst], len);
+        f->ifirst += len;
+        f->ilen -= len;
+        errno = 0;
+        return len;
+    }
+
+    /* Let's move all the data from the buffer first. */
+    char *pos = (char*)buf;
+    size_t remaining = len;
+    memcpy(pos, &f->ibuf[f->ifirst], f->ilen);
+    pos += f->ilen;
+    remaining -= f->ilen;
+    f->ifirst = 0;
+    f->ilen = 0;
+
+    mill_assert(remaining);
+    while(1) {
+        if(remaining > MILL_FILE_BUFLEN) {
+            /* If we still have a lot to read try to read it in one go directly
+             into the destination buffer. */
+            ssize_t sz = read(f->fd, pos, remaining);
+            if(!sz) {
+                return len - remaining;
+            }
+            if(sz == -1) {
+                if(errno != EAGAIN && errno != EWOULDBLOCK)
+                    return len - remaining;
+                sz = 0;
+            }
+            if((size_t)sz == remaining) {
+                errno = 0;
+                return len;
+            }
+            pos += sz;
+            remaining -= sz;
+            if (sz != 0 && fileeof(f)) {
+                return len - remaining;
+            }
+        }
+        else {
+            /* If we have just a little to read try to read the full connection
+             buffer to minimise the number of system calls. */
+            ssize_t sz = read(f->fd, f->ibuf, MILL_FILE_BUFLEN);
+            if(!sz) {
+                return len - remaining;
+            }
+            if(sz == -1) {
+                if(errno != EAGAIN && errno != EWOULDBLOCK)
+                    return len - remaining;
+                sz = 0;
+            }
+            if((size_t)sz < remaining) {
+                memcpy(pos, f->ibuf, sz);
+                pos += sz;
+                remaining -= sz;
+                f->ifirst = 0;
+                f->ilen = 0;
+            }
+            else {
+                memcpy(pos, f->ibuf, remaining);
+                f->ifirst = remaining;
+                f->ilen = sz - remaining;
+                errno = 0;
+                return len;
+            }
+            if (sz != 0 && fileeof(f)) {
+                return len - remaining;
+            }
+        }
+
+        /* Wait till there's more data to read. */
+        int res = fdwait(f->fd, FDW_IN, deadline);
+        if (!res) {
+            errno = ETIMEDOUT;
+            return len - remaining;
+        }
+    }
+}
+
+void fileclose(mfile f) {
+    fdclean(f->fd);
+    int rc = close(f->fd);
+    mill_assert(rc == 0);
+    free(f);
+    return;
+}
+
+mfile fileattach(int fd) {
+    struct mill_file *f = malloc(sizeof(struct mill_file));
+    if(!f) {
+        errno = ENOMEM;
+        return NULL;
+    }
+    int nfd = dup(fd);
+    mill_filetune(nfd);
```

Comment:
Yeah, while implementing this I noticed that this API can be used for other kinds of file descriptors, not only file descriptors that point to a "regular file". Sharing the implementation *internally* with other wrappers also *might* be ok.

Talking about "regular files" and having two file descriptors, one which is blocking and another one which isn't. My understanding from that stack overflow answer is that it is possible if you call `open(2)` twice, creating two file descript**ions** that point to the same file in disk. `dup(2)` doesn't open a new file descript**ion**, so I think it's just impossible to have a file descript**or** as input and have another file descript**or** as output pointing to a different file descript**ion** that points to the same underlying file inode. Unless we have a a way to get the `pathname` which was used with `open(2)` on the first file descript*or* so we can use it to call `open(2)` again. 

Some refefence:

http://man7.org/linux/man-pages/man2/open.2.html

> When a file descriptor is duplicated (using dup(2) or similar), the
       duplicate refers to the same open file description as the original
       file descriptor, and the two file descriptors consequently share the
       file offset and file status flags.  Such sharing can also occur
       between processes: a child process created via fork(2) inherits
       duplicates of its parent's file descriptors, and those duplicates
       refer to the same open file descriptions.

> Each open(2) of a file creates a new open file description; thus,
       there may be multiple open file descriptions corresponding to a file
       inode.

I don't know if this approach would apply to file descript*ors* that weren't created with `open(2)`. Either way it seems POSIX itself is broken as stated [here](http://homepage.ntlworld.com/jonathan.deboynepollard/FGA/dont-set-shared-file-descriptors-to-non-blocking-mode.html). I see two options, maybe we can try to find a way to duplicate a file descript*ion* and make it non blocking (maybe impossible). Or just make `mftune()`, `tcptune()`, etc, public so users take responsibility for tuning it or not.

Another thing, what would be the exact impact of having reads and writes on a file descriptor that is blocking on libmill? Other than the obvious blocking of other coroutines.